src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs

```diff
@@ -396,15 +396,15 @@ protected override void VisitSwitchSection(BoundSwitchSection node, bool isLastS
                             var tempState = this.State[tempSlot];
                             if (variableAccess is BoundLocal { LocalSymbol: SourceLocalSymbol local })
                             {
-                                var inferredType = TypeWithState.Create(tempType, tempState).ToTypeWithAnnotations();
+                                var value = TypeWithState.Create(tempType, tempState);
                                 if (_variableTypes.TryGetValue(local, out var existingType))
                                 {
                                     // merge inferred nullable annotation from different branches of the decision tree
-                                    _variableTypes[local] = TypeWithAnnotations.Create(existingType.Type, existingType.NullableAnnotation.Join(inferredType.NullableAnnotation));
+                                    _variableTypes[local] = TypeWithAnnotations.Create(existingType.Type, existingType.NullableAnnotation.Join(value.ToTypeWithAnnotations().NullableAnnotation));
                                 }
                                 else
                                 {
-                                    _variableTypes[local] = inferredType;
+                                    _variableTypes[local] = value.ToAnnotatedTypeWithAnnotations();
                                 }
```

Comment:
This should only be done for variables declared with a `var` pattern.  The others should stay non-null.